flight/Modules/VtolPathFollower/vtol_follower_fsm.c

```diff
@@ -255,18 +302,55 @@ const static struct vtol_fsm_transition *current_goal;
 static enum vtol_fsm_state curr_state;
 
 /**
- * Process any sequence of automatic state transitions
+ * Enter a new state.  Reacts appropriately (e.g. does nothing) if you pass
+ * FSM_STATE_UNCHANGED.
+ * @param[in] state The desired state.
+ * @return true if there was a state transition, false otherwise.
  */
-static void vtol_fsm_process_auto()
+static int vtol_fsm_enter_state(enum vtol_fsm_state state)
 {
-	while (current_goal[curr_state].next_state[FSM_EVENT_AUTO]) {
-		curr_state = current_goal[curr_state].next_state[FSM_EVENT_AUTO];
+	if (state != FSM_STATE_UNCHANGED) {
+		curr_state = state;
 
 		/* Call the entry function (if any) for the next state. */
 		if (current_goal[curr_state].entry_fn) {
 			current_goal[curr_state].entry_fn();
 		}
+
+		/* 0 disables any pending timeout, otherwise it's set to the
+		 * value for this state */
+		vtol_fsm_timer_set(current_goal[curr_state].timeout);
```

Comment:
doesn't matter for now, but later this might need to move up a few lines. that way a state can actually control it's timeout from the entry_fn() (imagine endpoints with programmable dwell times)